ffi/src/expressions.rs

```diff
@@ -245,3 +248,354 @@ pub extern "C" fn visit_expression_literal_bool(
 ) -> usize {
     wrap_expression(state, Expression::literal(value))
 }
+
+#[handle_descriptor(target=Expression, mutable=false, sized=true)]
+pub struct SharedExpression;
+
+/// Free the memory the passed SharedExpression
+///
+/// # Safety
+/// Engine is responsible for passing a valid SharedExpression
+#[no_mangle]
+pub unsafe extern "C" fn free_kernel_predicate(data: Handle<SharedExpression>) {
+    data.drop_handle();
+}
+
+/// The [`EngineExpressionVisitor`] defines a visitor system to allow engines to build their own
+/// representation of an expression from a particular expression within the kernel.
+///
+/// The model is list based. When the kernel needs a list, it will ask engine to allocate one of a
+/// particular size. Once allocated the engine returns an `id`, which can be any integer identifier
+/// ([`usize`]) the engine wants, and will be passed back to the engine to identify the list in the
+/// future.
+///
+/// Every expression the kernel visits belongs to some list of "sibling" elements. The schema
+/// itself is a list of schema elements, and every complex type (struct expression, array, variadic, etc)
+/// contains a list of "child" elements.
+///  1. Before visiting any complex expression type, the kernel asks the engine to allocate a list to
+///     hold its children
+///  2. When visiting any expression element, the kernel passes its parent's "child list" as the
+///     "sibling list" the element should be appended to:
+///      - For a struct literal, first visit each struct field and visit each value
+///      - For a struct expression, visit each sub expression.
+///      - For an array literal, visit each of the elements.
+///      - For a variadic `and` or `or` expression, visit each sub-expression.
+///      - For a binary operator expression, visit the left and right operands.
+///      - For a unary `is null` or `not` expression, visit the sub-expression.
+///  3. When visiting a complex expression, the kernel also passes the "child list" containing
+///     that element's (already-visited) children.
+///  4. The [`visit_expression`] method returns the id of the list of top-level columns
+///
+/// WARNING: The visitor MUST NOT retain internal references to string slices or binary data passed
+/// to visitor methods
+/// TODO: Add type information in struct field and null. This will likely involve using the schema visitor.
```

Comment:
I'd initially planned to add `StructField` visitor later on since it involved visiting a schema. But I guess we'll decide what to do with scalars first.